### DISCUSSION TITLE
Extract Data -> Restore Data SRR Pattern (no longer) possible in Nextjs?

### DISCUSSION THREAD

**Discussion by dgrahamlong** [urql-graphql/urql#2084] [Q&A] [answered]
Hi, just making sure this is possible because I could have sworn it used to work for me but now it's not. I got the idea from this video a while back so maybe something has changed since then. https://www.youtube.com/watch?v=Miock1yWkCQ&t=699s Bascially without messing around with next-urql, I'm doing this.

First I set up the client
```javascript
//Setting up and exporting client and ssrCache

const isClientBrowser: boolean = typeof window !== 'undefined'

export const ssrCache = ssrExchange({
  isClient: isClientBrowser,
});

const wsClient =
  isClientBrowser &&
  createWSClient({
    url: 'ws://localhost:3001/graphqlws',
  });

export const client = createClient({
  url: 'http://localhost:3001/graphql',
  fetchOptions: {
    credentials: 'include' as const,
    headers: {},
  },
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange({
      keys: {
        //keys
      },
      updates: {
        Mutation: {
          //Mutation Logic
        },
        Subscription: {
          //Subscription Logic
        },
      },
    }),
    ssrCache,
    fetchExchange,
    subscriptionExchange({
     //subscription exchange
    }),
  ],
});

```

In another page file I getServerSideProps and extract the data to urqlState

```javascript
import { client, ssrCache } from '@libs/urqlClient';

export async function getServerSideProps(ctx: NextPageContext) {

const res = await client
  .query(
    FETCH_FEED_POSTS,
    { limit: FETCH_LIMITS.feed },
    {
      fetchOptions: () => {
        //probably shouldn't be handling it this way but there's a cookie I want to pass in the headers
        return {
          headers: ctx.req?.headers,
        };
      },
    }
  )
  .toPromise();
console.log('RES', res);
return { props: { urqlState: ssrCache.extractData() } };
}

export default Feed


```

Finally in _app.js I restore the Data

```javascript
import { client, ssrCache } from '@libs/urqlClient';

export default function App({ Component, pageProps }: AppProps) {

  if (pageProps.urqlState) {
    console.log("URQL STATE", pageProps.urqlState)
    ssrCache.restoreData(pageProps.urqlState);
  }

  return (
          <Provider value={client}>
            <AppWrapper>
              <Layout>
                <Component {...pageProps} />
              </Layout>
            </AppWrapper>
          </Provider>
  );
}

```

The client side is still fetching data event though the "URQL STATE" console.log is showing the same data and the server is showing the same data and variables. Is it possible that there's two instances of the urql client?

**Comment by JoviDeCroock** [comment, marked as answer]
An issue with hydrating as a side-effect of render is that it sometimes comes late, hence we do it the following way in `next-urql` https://codesandbox.io/s/funny-lewin-do3b1?file=/lib/urql.js:324-341 I think some timings changed with the new version of Next 😅we've always done it this way to avoid these breakages

**Reply by dgrahamlong** [reply]
Ahh gotcha! That did it. Thanks for the help and the quick response!

**Comment by hyusetiawan** [comment]
@JoviDeCroock I have the same setup as @HummingBird24 and I am facing the same issue, same symptoms as well, where I can see the data in extract data has everything the app needs to render but still when it renders on the client-side, it still creates a graphql request.
From my debugging, it seems like ` ssrExchange.restoreData` does not actually restore the data even though I have confirmed that the data is there, it's lie the query does not read from it.

I checked out the link @JoviDeCroock  I added the initial render flag and all but still fetches on first render, are there other ideas I can try?

If it helps, here is what is being passed to restoreData
```
{
    "data": "{\"session\":{\"user\":{\"pid\":\"cl0q962xu0001q4acttpzyht8\",\"__typename\":\"User\",\"orgs\":[{\"org\":{\"pid\":\"cl0q962ya0017q4acjzxvasz8\",\"name\":\"ACME Inc.\",\"__typename\":\"Org\"},\"__typename\":\"OrgMembership\"}],\"name\":\"User1\",\"image\":null,\"email\":\"user1@gmail.com\"},\"__typename\":\"Session\"}}"
}
```

and this is what the query fetches during client runtime
```{
    "session": {
        "user": {
            "pid": "cl0q962xu0001q4acttpzyht8",
            "__typename": "User",
            "orgs": [
                {
                    "org": {
                        "pid": "cl0q962ya0017q4acjzxvasz8",
                        "name": "ACME Inc.",
                        "__typename": "Org"
                    },
                    "__typename": "OrgMembership"
                }
            ],
            "name": "User1",
            "image": null,
            "email": "user1@gmail.com"
        },
        "__typename": "Session"
    }
}
```
